### ISSUE TITLE
Panzer test fails on waterman

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#3372]
Building panzer on waterman fails to run 

How I built

```
source cmake/std/atdm/load-env.sh cuda-opt-Volta70
cmake -GNinja -DCMAKE_INSTALL_PREFIX=`pwd`/.. -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake -D Tpetra_ENABLE_TESTS=OFF -DAmesos2_ENABLE_Epetra=OFF -DTrilinos_ENABLE_Panzer=ON -DTrilinos_ENABLE_Phalanx=ON -DTrilinos_ENABLE_SEACASExodiff=ON -DTrilinos_ENABLE_SEACASEpu=ON -D Trilinos_ENABLE_SEACASAprepro:BOOL=ON -DMueLu_ENABLE_Kokkos_Refactor:BOOL=ON -DMueLu_ENABLE_Kokkos_Refactor_Use_By_Default:BOOL=ON -DXpetra_ENABLE_Kokkos_Refactor:BOOL=ON -DPanzer_ENABLE_TESTS=ON ~/workspace/Trilinos


ninja -j60

ctest -R PanzerAdaptersSTK_CurlLaplacianExample-ConvTest-Quad-Order-4
```

It fails in tpetra with a memory issue it looks like.  Trying to run debug code now..

@trilinos/panzer

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by bathmatt** [comment]
@rppawlo I'm building debug now, I will run it, see if it fails, run it through cuda-memcheck and hopefully get a traceback....

**Comment by bathmatt** [comment]
Ugh, cuda build times are 45+ minutes for Panzer_BasisValues2

**Comment by bathmatt** [comment]
cuda-memcheck doesn't really say much, there is an assertion thrown

:0: : block: [0,4,9], thread: [9,1,1] Assertion `Allocation failed.` failed.


then the traceback ,no other errors

**Comment by bathmatt** [comment]
Here is the last non-kokkos code


803	      Kokkos::parallel_for( policy, FunctorType(output, leftInput, rightInput, isTranspose) );

Here are the template defs..

#12 0x000000001372f948 in Intrepid2::ArrayTools<Kokkos::Cuda>::Internal::matvecProduct<Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda, Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda, Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda> (output=..., leftInput=..., rightInput=..., hasField=true, isTranspose=true)


The views that go into the functor look like this, 

```
(cuda-gdb) p output
$1 = {<Kokkos::ViewTraits<Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda>> = {<No data fields>}, m_track = {
    m_record_bits = 1, m_record = 0x1}, m_map = {m_handle = 0x2000c1400080, m_offset = {
      m_dim = {<Kokkos::Impl::ViewDimension0<0, 7>> = {N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 7>> = {
          N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 7>> = {N2 = 40}, <Kokkos::Impl::ViewDimension3<0, 7>> = {
          N3 = 2}, <Kokkos::Impl::ViewDimension4<0, 7>> = {N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 7>> = {
          N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 7>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<18446744073709551615, 7>> = {<No data fields>}, <No data fields>}, 
      m_stride = 16}, m_array_offset = {m_dim = {<Kokkos::Impl::ViewDimension0<0, 8>> = {
          N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 8>> = {N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 8>> = {
          N2 = 40}, <Kokkos::Impl::ViewDimension3<0, 8>> = {N3 = 2}, <Kokkos::Impl::ViewDimension4<0, 8>> = {
          N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 8>> = {N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 8>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<0, 8>> = {N7 = 41}, <No data fields>}, m_stride = 16}, m_fad_size = {
      value = 40}, m_fad_stride = {value = 51200}}, m_rank = 4}
(cuda-gdb) p leftInput
$2 = {<Kokkos::ViewTraits<Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda>> = {<No data fields>}, m_track = {
    m_record_bits = 1, m_record = 0x1}, m_map = {m_handle = 0x2000c1000080, m_offset = {
      m_dim = {<Kokkos::Impl::ViewDimension0<0, 7>> = {N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 7>> = {
          N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 7>> = {N2 = 2}, <Kokkos::Impl::ViewDimension3<0, 7>> = {
          N3 = 2}, <Kokkos::Impl::ViewDimension4<0, 7>> = {N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 7>> = {
          N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 7>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<18446744073709551615, 7>> = {<No data fields>}, <No data fields>}, 
      m_stride = 16}, m_array_offset = {m_dim = {<Kokkos::Impl::ViewDimension0<0, 8>> = {
          N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 8>> = {N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 8>> = {
          N2 = 2}, <Kokkos::Impl::ViewDimension3<0, 8>> = {N3 = 2}, <Kokkos::Impl::ViewDimension4<0, 8>> = {
          N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 8>> = {N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 8>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<0, 8>> = {N7 = 41}, <No data fields>}, m_stride = 16}, m_fad_size = {
      value = 40}, m_fad_stride = {value = 2560}}, m_rank = 4}
(cuda-gdb) p rightInput
$3 = {<Kokkos::ViewTraits<Sacado::Fad::DFad<double>, Kokkos::LayoutLeft, Kokkos::Cuda>> = {<No data fields>}, m_track = {
    m_record_bits = 1, m_record = 0x1}, m_map = {m_handle = 0x2000c1200080, m_offset = {
      m_dim = {<Kokkos::Impl::ViewDimension0<0, 7>> = {N0 = 40}, <Kokkos::Impl::ViewDimension1<0, 7>> = {
          N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 7>> = {N2 = 2}, <Kokkos::Impl::ViewDimension3<0, 7>> = {
          N3 = 1}, <Kokkos::Impl::ViewDimension4<0, 7>> = {N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 7>> = {
          N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 7>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<18446744073709551615, 7>> = {<No data fields>}, <No data fields>}, 
      m_stride = 40}, m_array_offset = {m_dim = {<Kokkos::Impl::ViewDimension0<0, 8>> = {
          N0 = 40}, <Kokkos::Impl::ViewDimension1<0, 8>> = {N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 8>> = {
          N2 = 2}, <Kokkos::Impl::ViewDimension3<0, 8>> = {N3 = 1}, <Kokkos::Impl::ViewDimension4<0, 8>> = {
          N4 = 1}, <Kokkos::Impl::ViewDimension5<0, 8>> = {N5 = 1}, <Kokkos::Impl::ViewDimension6<0, 8>> = {
          N6 = 1}, <Kokkos::Impl::ViewDimension7<0, 8>> = {N7 = 41}, <No data fields>}, m_stride = 40}, m_fad_size = {
      value = 40}, m_fad_stride = {value = 3200}}, m_rank = 3}


```

@crtrott do you see anything wrong with these??

**Comment by rppawlo** [comment]
@bathmatt - we've seen this when the device runs out of memory. This test is at the high end of a convergence test. In the past it was a result from all mpi processes on a node hitting the same gpu card. I can take a look at this next week.

**Comment by rppawlo** [comment]
> Ugh, cuda build times are 45+ minutes for Panzer_BasisValues2

yeah - I noticed this too. I've got a potential fix this but need to make some changes to both panzer and EMPIRE. Let's talk about this next week.

**Comment by bathmatt** [comment]
> @bathmatt - we've seen this when the device runs out of memory. This test is at the high end of a convergence test. In the past it was a result from all mpi processes on a node hitting the same gpu card. I can take a look at this next week.

This is a 2D 16x16 test.  4th order but this works on smaller cards,the volta has 16G each

**Comment by bathmatt** [comment]
Actually, this only fails in parallel, so I don't think it is running out of memory..

**Comment by rppawlo** [comment]
Looks like the sacado derivative dimension might not be initialized. Maybe it is some large random number, killing memory.

**Comment by mperego** [comment]
I agree with @rppawlo . @bathmatt  can you check if the views sent to `Intrepid2::FunctionSpaceTools<PHX::Device::execution_space>::
      HCURLtransformVALUE(basis_vector.get_view(),
                          jac_inv.get_view(),
                          basis_ref_vector.get_view());`
here /ascldap/users/mbetten/workspace/Trilinos/packages/panzer/disc-fe/src/Panzer_BasisValues2.cpp:104
have meaningful derivative dimensions?

**Comment by bathmatt** [comment]
This is what panzer passes in..

The 4 real dims seem right, the fad dim seems wacked..  
```
108         
(gdb) p basis_vector
$1 = {static ArrayRank = <optimized out>, m_tag = {ptr_ = 0x60882ac0, node_ = {node_ = 0x60477de0, strength_ = Teuchos::RCP_STRONG}}, 
  m_field_data = {<Kokkos::ViewTraits<Sacado::Fad::DFad<double>****, Kokkos::LayoutLeft, Kokkos::Cuda>> = {<No data fields>}, m_track = {m_record_bits = 1589584576, m_record = 0x5ebf22c0}, m_map = {m_handle = 0x2000c1400080, 
      m_offset = {m_dim = {<Kokkos::Impl::ViewDimension0<0, 4>> = {N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 4>> = {N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 4>> = {N2 = 40}, <Kokkos::Impl::ViewDimension3<0, 4>> = {
            N3 = 2}, <Kokkos::Impl::ViewDimension4<18446744073709551615, 4>> = {<No data fields>}, <Kokkos::Impl::ViewDimension5<18446744073709551615, 4>> = {<No data fields>}, <Kokkos::Impl::ViewDimension6<18446744073709551615, 4>> = {<No data fields>}, <Kokkos::Impl::ViewDimension7<18446744073709551615, 4>> = {<No data fields>}, <No data fields>}, m_stride = 16}, m_array_offset = {m_dim = {<Kokkos::Impl::ViewDimension0<0, 5>> = {
            N0 = 16}, <Kokkos::Impl::ViewDimension1<0, 5>> = {N1 = 40}, <Kokkos::Impl::ViewDimension2<0, 5>> = {N2 = 40}, <Kokkos::Impl::ViewDimension3<0, 5>> = {N3 = 2}, <Kokkos::Impl::ViewDimension4<0, 5>> = {
            N4 = 41}, <Kokkos::Impl::ViewDimension5<18446744073709551615, 5>> = {<No data fields>}, <Kokkos::Impl::ViewDimension6<18446744073709551615, 5>> = {<No data fields>}, <Kokkos::Impl::ViewDimension7<18446744073709551615, 5>> = {<No data fields>}, <No data fields>}, m_stride = 16}, m_fad_size = {value = 40}, m_fad_stride = {value = 51200}}}}
(gdb) 


```

**Comment by mperego** [comment]
@bathmatt , the fad dim seems right to me, it's 41 (40 derivatives + 1 for value) which is the same as the cardinality of the basis. 
I looked at this with @ndellingwood and he doesn't see anything obviously wrong. The number `18446744073709551615` is the KOKKOS_INVALID_INDEX, and it's likely due to the fact that we are using dynamic Fad type. The output of the dynamic views in Intrepid2 is a bit funky, but again it's not clear to us whether it's wrong.

**Comment by bartlettroscoe** [comment]
@bathmatt, is this a duplicate of #3340?

**Comment by bartlettroscoe** [comment]
Closing as a duplicate of earlier issue #3340.
